Initial design for ExternalService API with service for public IP

RaitoBezarius opened the pull request:
<!---
    Thank you for contributing to Aqua Security.
    Please don't remove the template.
-->

## Description
As per to #235 ; I'm trying to see what kind of external service API would fit the usage in Kube Hunter (as in, it's generalizable for other external services).
Here is an initial idea, I didn't want to go full OO magic, using singletons pattern & whatever.

This is a very simple PR, though it has the potential for external plugins to use their own providers of public IP, it can be used to modify if you want IPv4/IPv6, it can accommodate special use cases: when you want *private* IP, let's say, you're in a private VPC or something like this.

Two things are left as questions and require feedback:

- Test, how do we want to go? I can test for generic failures using mock and ensure that `try_get` calls the second provider. That's a simple enough and relevant test to do. But before, I want to know if that'd be interesting to move this logic into a base class called `ServiceWithFailover` which enables this pattern.
- Propagate a form of configuration to the modules, if we want to globally use only this provider, we may want that all subsequent instances of ExternalService instantiates using only those specific providers, what I usually suggest is to use module name which are importable: `["plugins.external_providers.canhazip", "plugins.external_providers.ipify"]` for example, this can be configured in some file or even in the `conf.py` if #264 is merged.

This enables new patterns to configure stuff using environment variables too. Anyway, I need feedback to understand what maintainers feel is the most appropriate for this project.

Bonus: Ipify.org is added.

## Fixed Issues

Fixes #235 

## Contribution checklist
 - [x] I have read the Contributing Guidelines.
 - [x] The commits refer to an active issue in the repository.
 - [ ] I have added automated testing to cover this case.
 
## Notes
As above-mentioned, automated testing is a question.

lizrice commented:
Sorry it has taken so long to get back to you on this @RaitoBezarius - we had a lot of Hacktoberfest changes and then KubeCon. I like the general idea of this, will leave it with @iyehuda and @itaysk to figure out how they want to take this forward. 

I realized we could be querying http**s**://canihazip.com/**s** which returns simply the IP address, which I think would be better than http://canihazip.com. Which then begs the question whether it's worth falling back between one service and another identical one, but I suppose one of them might be down.

RaitoBezarius commented:
@lizrice No problem, I have some cleanup to do myself on my contributions. Regarding the HTTPS vs HTTP stuff, it's baked in this PR (except the final `/s` I suppose), this was a low hanging fruit I guess which is cool.
One of them might be down, worst, one of them might be blocked (?).

To me, as a red teamer, this is the kind of the tool I'd use for initial recognition (it's super cool and I really appreciate Python was used for this because it makes it easier for me to extend than Ruby/Perl stuff I deal with), I have done some tests with it and it gave really interesting results.

In a corporate environment or in certain super strange contexts, it is really important for me to control that kind of service.
Better, it enables me to switch a public IP service for a private IP service, which is really neat in cloud infrastructures (if it makes sense for the tool or during an assessment).

So… I guess I'll just do a real effort and clean my PRs so you guys can see how do you want to move forward with this.

iYehuda commented on src/modules/external_services/external_ip.py:
```diff
@@ -0,0 +1,45 @@
+    except requests.ConnectionError as e:
+        raise ExternalProviderFailure(e)
+
+ExtIPProvider = namedtuple('ExternalIPProvider', [
```
Rename to `ExternalIPProvider`

iYehuda commented on src/modules/external_services/external_ip.py:
```diff
@@ -0,0 +1,45 @@
+        self.providers = sorted(providers or DEFAULT_PROVIDERS, key=lambda provider: provider.priority, reverse=True)
+        self.cache_results = cache_results
+
+    def try_get(self, bypass_cache=False):
```
I think it is better to have `use_cache` parameter with default value of `True` since you check if cache is wanted

iYehuda commented on src/modules/external_services/exceptions.py:
```diff
@@ -0,0 +1,7 @@
+
+class ExternalProviderFailure(Exception):
+    def __init__(self, exc=None):
+        self.wrapped_exc = exc
```
The difference between these exceptions is not clear (I understood that after reading the using code).
Please add docstring that will clarify the usage and difference

iYehuda commented on src/modules/external_services/external_ip.py:
```diff
@@ -0,0 +1,45 @@
+        self.cache_results = cache_results
+
+    def try_get(self, bypass_cache=False):
+        if not bypass_cache and self.cache_results and self.cached_ip:
```
`self.cached_ip` should be defined in `__init__`

iYehuda commented on src/modules/external_services/external_ip.py:
```diff
@@ -0,0 +1,45 @@
+        raise ExternalProviderFailure(e)
+
+ExtIPProvider = namedtuple('ExternalIPProvider', [
+    'handler',
```
Use more imperative name (e.g. `get`/`fetch`/`provide`) to make the usage more clear

iYehuda commented on src/modules/external_services/external_ip.py:
```diff
@@ -0,0 +1,45 @@
+        for provider in self.providers:
+            try:
+                ip = provider.handler()
+                self.cached_ip = ip
```
Cache the result if `self.cache_results` is `True`
It has no major effect with IP address string but it can get more memory intensive with other data services.

iYehuda commented on src/modules/external_services/external_ip.py:
```diff
@@ -0,0 +1,45 @@
+
+DEFAULT_PROVIDERS = [CanHazIP, Ipify]
+
+class ExternalIPService:
```
I think this class can be a general `BaseExternalService` with failover in it so it can be reused

RaitoBezarius replied:
Do you mean like, I should go ahead and generalize it for any type of results?

RaitoBezarius replied:
I felt like I could do it, but I was wondering if it was a really good idea, but if you want I can do the modifications and make it a `external_base_service.py` and derive `external_ip.py` from it.

iYehuda replied:
It feels right to me, I think it will save some code duplication in future.
Let's go with a `base_external_service.py` with `BaseExternalService` class.

RaitoBezarius replied:
Alright, I'll perform those last changes tomorrow then and fix conflicts

iYehuda commented:
Hi @RaitoBezarius !

Any progress made?

RaitoBezarius commented:
> Hi @RaitoBezarius !
> 
> Any progress made?

Hi there, I addressed most of the comments, I'll fix the conflicts and I need to ask you regarding the base class stuff.

iYehuda commented on src/modules/external_services/external_ip.py:
```diff
@@ -0,0 +1,51 @@
+])
+
+CanHazIP = ExternalIPProvider(provide=generic_text_http_provider('https://canhazip.com'), priority=0)
+Ipify = ExternalIPProvider(provide=generic_text_http_provider('https://api.ipify.org'), priority=1)
```
I don't think using priority fits here.
In my opinion, the service provider mechanism should be as easy as possible to use.
Maybe just use a list of them so first item is first priority?

RaitoBezarius replied:
Sure, but if some user want to dynamically modify the providers and want to have his own to be used rather than the system ones, I'm not sure how it's going to be done.

iYehuda replied:
In that case we can build the failover list with his preference first and append the original list, filtered:
```python
failover_list = [preference] + [service for service in services if service != preference]
```

danielsagi commented:
Closing.. a lot of those changes are not longer relevant. changing cloud discovery is  already planned to be implemented differently. and we no longer rely on canhazip.com
